concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+between null and an empty string, or the difference between null and absent?
+This is why we keep counting cardinality even after an infinity shows up!)
+
+### unit types
+
+Let's talk about one last size of cardinality... "unit": unit just means "one".
+
+"Unit" types don't come up too often... or at least, they're not often named as such.
+Just like "product types" are quite common once you recognize them, though, so is "unit".
+
+"Null" can be thought of as a "unit" type.
+
+"Maybe"/"Optional" types (further discussion later in the [Types in the Wild](#types-in-the-wild) chapter,
+specifically the [maybe types](#maybe-types) section) can be thought of as "unit plus a thing".
+
+Enums can be considered a sum type of where every type is itself a just another named unit type!
+(Whether or not you find this a *useful* perspective is up to you.  If it's confusing, forget it.)
+
+
+Open vs Closed membership
+-------------------------
+
+All of the above [categories of types](#categories-of-types) have one thing in common:
+they have "closed" membership.  That is, at "compile" (or other static analysis) time,
+we know what their membership is, and we can completely count their cardinality (infinities notwithstanding).
+Once the program is compiled(/analyzed), there's no way to add new members to any of these types,
+and no reason to consider the idea of updating any of the cardinality accounting.
+
+What if we want "open" membership, where I *don't* know all of the possible
+types of values that will be part of a type in advance?
+
+This "open membership" is commonly known as **interfaces**.
+
+### interfaces
+
+Interfaces have the concept of "open membership", and are also typically
+combined with some definition of "features" that are required if something
+will be considered to implement the interface.
+
+In programming languages, the definition of "features" usually is expressed in
+terms of a set of methods.  This makes sense for programming languages,
+because if you don't know what methods you can call on a thing,
+you can't really do much with the thing!
+
+One could also imagine using a different definition of "features" to describe
+"interfaces" that apply purely to data:
+For example, stating that some map has the features to conform to an interface
+only if some entries of some known names and value types are present in the map.
+(However, this is not often seen, so this document will continue to frequently
+refer to "sets of methods" when discussing interfaces.)
+
+### explicit versus structural interfaces
+
+Interfaces as found in programming languages tend to fall into one of two large camps:
+*explicit* interfaces, or *structural* interfaces.
+
+Explicit interfaces require types implementing them to say so.
+
+Structural interfaces state some set of methods (or more generally, features)
+that must be present on a type for it to satisfy the interface;
+then, any type that has those methods implements the interface.
```

Comment:
I.e. "duck typing"?

Reply:
I tend to think of them as different, but I outright resorted to plugging "structural typing vs duck typing" into a search engine, because this does seem murky and arguable to me.  One of the first things that came up was https://softwareengineering.stackexchange.com/questions/259943/is-there-a-difference-between-duck-typing-and-structural-typing , and I think I more or less agree with the answer there.

"Duck typing" tends to refer to probing a thing (and typically, at _run_ time, rather than compile time) to see if it can do what you want.  I also tend to conceptualize it as probing for *specifically* the function you're about to call -- this is often what it means in practice when discussed in the Ruby community (and if they didn't originate this terminology, they're certainly at least where I'm familiar with it from).

"Structural typing" involves having a whole suite of constraints in a bundle, and checking them all at once.  It also seems to be the phrasing that's used more often when people are discussing compile time checks.

But I'm using a lot of words like "tends" and "seems" and "often".  So I agree they're pretty dang similar.

I'm also just plain biased here because I've seen "duck typing" used to justify a lot of bad code over the years, and for whatever reason, I've seen very little of the same for "structural typing".  Does that seem nonsensical, given that they're describing almost the same thing, when one dissects it?  Yes.  Do I make this observation nonetheless?  Also yes.

So if I have to choose one term to signal-boost: it's gonna be "structural typing".  And the distinction is murky enough I'm not actually sure I want to draw attention to it in this primer.